### PR TITLE
Close; Anonymise; Erase

### PR DESCRIPTION
**Pull request by garethrees**
Add some task-specific methods for better composability and future ability to add a UI button for each action to give us more control over how we handle various user requests.

A precursor to:

* https://github.com/mysociety/alaveteli/issues/285
* https://github.com/mysociety/alaveteli/issues/5052
* https://github.com/mysociety/alaveteli/issues/4922
* https://github.com/mysociety/alaveteli/issues/4910

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -416,6 +416,31 @@ def closed?
...
   end
 
+  def erase
+    erase!
+  rescue ActiveRecord::RecordInvalid, ArgumentError
+    false
+  end
+
+  def erase!
+    raise ArgumentError unless closed?
```
Comment (gbp) [inline app/models/user.rb:426]: Not sure `ArgumentError` is the best exception to raise here. Raising `ActiveRecord::RecordInvalid` or our own `RecordInvalid` exception would be better imo

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -1246,6 +1246,112 @@ def create_user(options = {})
...
   end
 
+  describe '#erase' do
+    subject { user.erase }
+
+    let(:user) { FactoryBot.build(:user) }
+
+    context 'the update is successful' do
+      before do
+        user.close!
```
Comment (gbp) [inline spec/models/user_spec.rb:1257]: Should we create a `closed` factory trait?

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -1246,6 +1246,112 @@ def create_user(options = {})
...
+        expect(user).to receive(:erase!).and_call_original
+        subject
+      end
+
+      it { is_expected.to eq(true) }
+
+      it 'erases the account' do
+        expect(user.name).to match(/Name Removed/)
```
Comment (gbp) [inline spec/models/user_spec.rb:1265]: Should this match line 1315?

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -1191,6 +1247,138 @@ def create_user(options = {})
...
+    context 'the update is unsuccessful' do
+      before do
+        expect(user).to receive(:erase!).and_raise(ActiveRecord::RecordInvalid)
+        subject
+      end
+
+      it { is_expected.to eq(false) }
+
+      it 'does not erase the account' do
+        expect(user.name).not_to match(/Name Removed/)
```
Comment (gbp) [inline spec/models/user_spec.rb:1278]: See above

**Review by gbp — APPROVED** [review]
Looks good. Left some minor comments.

File: spec/models/user_spec.rb (hunk 2 above)
Comment (garethrees) [reply spec/models/user_spec.rb:1257]: Ah, already had one! Updated.

**Pull request merged**